# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/arm64/CodeGenerator-arm64.cpp
# Commit: 27db0e6a7010
# Full Hash: 27db0e6a7010f290ac14e5d260c0b4fe48442486
# Author: Lars T Hansen <lhansen@mozilla.com>
# Date: 2021-10-18 16:09:29
# Regressor Bug: 1727084
# File Overlap Count: 6
# Description:
#   Bug 1727084 - Memory64 - Ion operations. r=yury
#   
#   Add memory64 generalizations for Ion for all operations (except bulk
#   memory, done in an earlier patch).
#   
# ==============================================================================

diff -r 1ab8383d17ee -r 27db0e6a7010 js/src/jit/arm64/CodeGenerator-arm64.cpp
--- a/js/src/jit/arm64/CodeGenerator-arm64.cpp	Mon Oct 18 07:31:56 2021 +0000
+++ b/js/src/jit/arm64/CodeGenerator-arm64.cpp	Mon Oct 18 07:31:56 2021 +0000
@@ -2295,8 +2295,11 @@
 
 static Maybe<uint64_t> IsAbsoluteAddress(const LAllocation* ptr,
                                          const wasm::MemoryAccessDesc& access) {
-  if (ptr->isConstant()) {
-    uint64_t base_address = uint32_t(ToInt32(ptr));
+  if (ptr->isConstantValue()) {
+    const MConstant* c = ptr->toConstant();
+    uint64_t base_address = c->type() == MIRType::Int32
+                                ? uint64_t(uint32_t(c->toInt32()))
+                                : uint64_t(c->toInt64());
     uint64_t offset = access.offset();
     return Some(base_address + offset);
   }
@@ -2312,6 +2315,8 @@
     return;
   }
 
+  // ptr is a GPR and is either a 32-bit value zero-extended to 64-bit, or a
+  // true 64-bit value.
   masm.wasmLoad(mir->access(), HeapReg, ToRegister(lir->ptr()),
                 ToAnyRegister(lir->output()));
 }
@@ -2558,6 +2563,20 @@
   masm.bind(&ok);
 }
 
+void CodeGenerator::visitWasmAddOffset64(LWasmAddOffset64* lir) {
+  MWasmAddOffset* mir = lir->mir();
+  Register64 base = ToRegister64(lir->base());
+  Register64 out = ToOutRegister64(lir);
+
+  masm.Adds(ARMRegister(out.reg, 64), ARMRegister(base.reg, 64),
+            Operand(mir->offset()));
+
+  Label ok;
+  masm.j(Assembler::CarryClear, &ok);
+  masm.wasmTrap(wasm::Trap::OutOfBounds, mir->bytecodeOffset());
+  masm.bind(&ok);
+}
+
 void CodeGenerator::visitWasmSelectI64(LWasmSelectI64* lir) {
   MOZ_ASSERT(lir->mir()->type() == MIRType::Int64);
   Register condReg = ToRegister(lir->condExpr());